mutable_tree.go

```diff
@@ -1028,53 +1028,57 @@ func (tree *MutableTree) balance(node *Node) (newSelf *Node, err error) {
 // saveNewNodes save new created nodes by the changes of the working tree.
 // NOTE: This function clears leftNode/rigthNode recursively and
 // calls _hash() on the given node.
+// TODO: Come back and figure out how to better parallelize this code.
 func (tree *MutableTree) saveNewNodes(version int64) error {
 	nonce := uint32(0)
-	newNodes := make([]*Node, 0)
-	var recursiveAssignKey func(*Node) ([]byte, error)
-	recursiveAssignKey = func(node *Node) ([]byte, error) {
+
+	nodeChan := make(chan *Node, 64) // Buffered channel to avoid blocking.
+	doneChan := make(chan error, 1)  // Channel to signal completion and return any errors.
+
+	// Start a goroutine to save nodes.
+	go func() {
+		var err error
+		for node := range nodeChan {
+			if saveErr := tree.ndb.SaveNode(node); saveErr != nil {
+				err = saveErr
+				break
+			}
+			node.leftNode, node.rightNode = nil, nil // Detach children after saving.
+		}
+		doneChan <- err // Send any error encountered or nil if none.
+	}()
+
+	var recursiveAssignKey func(*Node) []byte
+	recursiveAssignKey = func(node *Node) []byte {
 		if node.nodeKey != nil {
 			if node.nodeKey.nonce != 0 {
-				return node.nodeKey.GetKey(), nil
+				return node.nodeKey.GetKey()
 			}
-			return node.hash, nil
+			return node.hash
 		}
 		nonce++
 		node.nodeKey = &NodeKey{
 			version: version,
-			nonce:   nonce,
+			nonce:   nonce, // Example nonce calculation; adjust as needed.
```

Comment:
just realized the original codebase is wrong, not a bug but just a performance issue.
`newNodes = append(newNodes, node)` should come after this line, the main idea of `newNodes` slice is to save the node in a sorted manner by the nodekey (here the nonce) to reduce the compaction.

Just worried we can't assume this sorted manner after refactoring ...